spaghetto/src/Spago/Command/Ls.purs

```diff
@@ -1,3 +1,83 @@
-module Spago.Command.Ls where
+module Spago.Command.Ls (listPackages, listPackageSet, LsEnv(..), LsDepsArgs, LsPackagesArgs) where
 
-import Prelude
+import Spago.Prelude
+
+import Data.Foldable (elem)
+import Data.Map (filterKeys)
+import Data.Map as Map
+import Data.Set (unions)
+import Data.Tuple.Nested (type (/\))
+import Registry.Internal.Codec (packageMap)
+import Registry.PackageName as PackageName
+import Registry.Version as Version
+import Spago.Config (Dependencies(..), Package(..), PackageSet, Workspace, packageCodec)
+import Spago.Config as Config
+
+type LsPackagesArgs =
+  { json :: Boolean
+  }
+
+type LsDepsArgs =
+  { json :: Boolean
+  , transitive :: Boolean
+  }
+
+type LsEnv =
+  { dependencies :: PackageSet
+  , logOptions :: LogOptions
+  , workspace :: Workspace
+  }
+
+listPackageSet :: LsPackagesArgs -> Spago LsEnv Unit
+listPackageSet { json } = do
+  logDebug "Running `listPackageSet`"
+  { workspace } <- ask
+  output $ formatPackageNames json (Map.toUnfoldable workspace.packageSet)
+
+listPackages :: LsDepsArgs -> Spago LsEnv Unit
+listPackages { transitive, json } = do
+  logDebug "Running `listPackages`"
+  { dependencies, workspace } <- ask
+  let
+    directDependencies = filterKeys (_ `elem` direct) dependencies
+      where
+      toDependencyNames (Dependencies deps) = Map.keys deps
+      packagesToNames = toDependencyNames <<< _.dependencies <<< _.package
+      direct = unions (packagesToNames <$> workspacePackages)
+      workspacePackages = case workspace.selected of
+        Just p -> [ p ]
+        Nothing -> Config.getWorkspacePackages workspace.packageSet
```

Comment:
Then this gets easier once we have only one package (it can be inlined more, but this is just to give an idea):
```suggestion
  { dependencies, workspace, selected } <- ask
  let
    directDependencies = filterKeys (_ `elem` direct) dependencies
      where
      toDependencyNames (Dependencies deps) = Map.keys deps
      packagesToNames = toDependencyNames <<< _.dependencies <<< _.package
      direct = packagesToNames selected
```